### PR TITLE
Feat/add modal window

### PR DESCRIPTION
**Pull request by craftez**
I tried to make the commits as clear as possible.
1. Remove the index test because it was broken after the refactor of multi-stacking modals.
2. I tried to built it as flexible as possible, so it could be extended or adjusted to needs.
3. THere is a small issue to address when multi-stacking modals, which is the overlay been added as part of the modal, but i believe as an mvp of the required features i got most of it cover.
4. I consider accessibility and flexibility as the core feature of this modal. 
5. Also used a compound pattern to add modal content, and the mechanism for stacking is pretty simple, but it gives an idea of where to go next.

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -18,17 +18,25 @@
     "react-dom": "18.2.0"
   },
   "devDependencies": {
+    "@testing-library/jest-dom": "^5.16.5",
+    "@testing-library/react": "^14.0.0",
+    "@testing-library/react-hooks": "^8.0.1",
+    "@testing-library/user-event": "^14.4.3",
```
Comment (craftez) [inline frontend/package.json:24]: Deps for unit tests.

File: frontend/package.json [hunk 2]
```diff
@@ -18,17 +18,25 @@
...
     "eslint": "8.18.0",
     "eslint-config-canonical": "35.0.1",
     "eslint-config-next": "12.1.6",
     "eslint-config-prettier": "8.5.0",
     "jest": "28.1.1",
+    "jest-environment-jsdom": "^29.5.0",
     "lint-staged": "13.0.3",
+    "postcss": "^8.4.23",
     "prettier": "2.7.1",
+    "tailwindcss": "^3.3.2",
```
Comment (craftez) [inline frontend/package.json:39]: Decided to use tailwindcss as i assumed the focus was more on the modal js code more than how it looks.

File: frontend/src/components/Modal.tsx [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  onClose: () => void;
+};
+
+export const Modal: React.FC<ModalProps> = ({
+  id,
+  isOpen,
+  onClose,
+  children,
+}) => {
+  // eslint-disable-next-line react/hook-use-state
```
Comment (craftez) [inline frontend/src/components/Modal.tsx:19]: Given that i was on a rush, and was not sure how much i should respect the original configs lint rules i had to disable some rules here and there, i know it should be clean.... but i was trying to get the code out. :)

File: frontend/src/components/Modal.tsx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+  useEffect(() => {
+    // eslint-disable-next-line no-debugger
+    if (isOpen) {
+      pushModal(id);
+    } else {
+      popModal();
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [isOpen, id]);
```
Comment (craftez) [inline frontend/src/components/Modal.tsx:31]: here there are some pending deps, but the react render cycle was getting out of hand, so i had to limit it to what was relevant, again i think some strict rules should be ease in this kinda of case, given that we should add direct deps

File: frontend/src/components/Modal.tsx [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      <div
+        className="w-full h-full flex items-center justify-center"
+        onClick={(event) => {
+          if (
+            event.target === modalRef.current &&
+            modalStack[modalStack.length - 1] === id
+          ) {
+            onClose();
+          }
+        }}
```
Comment (craftez) [inline frontend/src/components/Modal.tsx:53]: Seen this, i would def. move this out... trying to keep it clean and readable.

File: frontend/src/components/Modal.tsx [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  }, [isOpen, id]);
+
+  if (!isOpen) return null;
+
+  return ReactDOM.createPortal(
+    <div
+      aria-modal="true"
+      className="fixed inset-0 bg-neutral-800 bg-opacity-50 "
+      ref={modalRef}
+      role="dialog"
```
Comment (craftez) [inline frontend/src/components/Modal.tsx:40]: use this role here for more semantic represenation and accesibility, the inner div with presentation role was added given the linting rules.

File: frontend/src/hooks/useModal.tsx [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import { useEffect, useRef } from 'react';
+
```
Comment (craftez) [inline frontend/src/hooks/useModal.tsx:2]: Core file, this is the modal mechanism. It tracks and handle events such as escaping. And then then also scroll-locking ;).

Comment (craftez) [reply frontend/src/hooks/useModal.tsx:2]: and i forgot focus.

File: frontend/src/pages/index.tsx [hunk 8]
```diff
@@ -1,8 +1,66 @@
 /* eslint-disable canonical/filename-match-exported */
 import { type NextPage } from 'next';
+import { useState } from 'react';
+import { Modal } from '@/components/Modal';
 
 const Index: NextPage = () => {
-  return <h1>Welcome to Contra!</h1>;
+  const [modalAOpen, setModalAOpen] = useState(false);
```
Comment (craftez) [inline frontend/src/pages/index.tsx:7]: on this file i was trying to make it so we were able to open a modal form within a modal, (which is not such a good UX practice), but the goal was to test couple of edge cases. I also had buttons side by side opening or closing modals.
